### DISCUSSION TITLE
Question about function style

### DISCUSSION THREAD

**Discussion by Rudxain** [darkreader/darkreader#10394] [Q&A] [answered]
I haven't read all JS/TS files in the repo, but I don't understand the "rules" or "pattern" about "classic"-fns and arrow-fns. At first, I thought "ok, they use `function` for subroutines, and `=>` for expressions", but then I found some files where `=>` is also used for subroutines, but only within the scope of a "classic"-fn.

I couldn't find ESlint rules for this, so I'm kinda lost

**Comment by Gusted** [comment, marked as answer]
> I haven't read all JS/TS files in the repo, but I don't understand the "rules" or "pattern" about "classic"-fns and arrow-fns. At first, I thought "ok, they use `function` for subroutines, and `=>` for expressions", but then I found some files where `=>` is also used for subroutines, but only within the scope of a "classic"-fn.
> 
> I couldn't find ESlint rules for this, so I'm kinda lost

Are you trying to accomplish something? I would recommend that you would [read about function expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions#the_function_expression) in javascript. There isn't a ESLint for this, as choosing which to use depends on the situation, e.g. for passing a function or one-liners you would use anonymous functions, but for functions that contain a lot of code you would go with the standard expression.

**Reply by Rudxain** [reply]
I understand. What I mean is that I've seen `function` declarations and "arrow declarations" (I know these don't exist, I mean stuff like `const f = x => x;`) used almost interchangeably, even when `this` isn't used at all.

I personally prefer to use arrows everywhere, and only use "classic" fns when I'm concerned about `this`.

Arrow fns can have automatic `name` properties (only when used directly in a variable declaration), meaning that debugging stack traces is as easy as with named `function`s

Another advantage is that `const` fns are **way shorter** to declare with `=>` than with `function`, even when `function` is anonymous. Function declarations are mutable by default, which is usually an undesired thing

**Reply by Gusted** [reply]
I'm sorry, but I'm losing the relevance to Dark Reader here...

**Reply by Rudxain** [reply]
It's for contributions! I want to know, just-in-case I open a PR adding some code (in the future, not now)

**Reply by Gusted** [reply]
Aha I see. Well `function` is the way to go, it's what you get learned, what everyone is used to and notably, it's already written like that in the code.  I personally think the code style and readability goes down if you use too much arrow notations.

**Reply by Rudxain** [reply]
Oh ok. Thanks for explaining!
